hbase-server/src/main/java/org/apache/hadoop/hbase/master/RegionsRecoveryChore.java

```diff
@@ -0,0 +1,154 @@
+
+  private static final String ERROR_REOPEN_REIONS_MSG =
+    "Error reopening regions with high storeRefCount. ";
+
+  private final HMaster hMaster;
+  private final int storeRefCountThreshold;
+
+  private static final PerClientRandomNonceGenerator NONCE_GENERATOR =
+    PerClientRandomNonceGenerator.get();
+
+  /**
+   * Construct RegionsRecoveryChore with provided params
+   *
+   * @param stopper When {@link Stoppable#isStopped()} is true, this chore will cancel and cleanup
+   * @param configuration The configuration params to be used
+   * @param hMaster HMaster instance to initiate RegionTableRegions
+   */
+  RegionsRecoveryChore(final Stoppable stopper, final Configuration configuration,
+      final HMaster hMaster) {
+
+    super(REGIONS_RECOVERY_CHORE_NAME, stopper, configuration.getInt(REGIONS_RECOVERY_INTERVAL,
+      DEFAULT_REGIONS_RECOVERY_INTERVAL));
+    this.hMaster = hMaster;
+    this.storeRefCountThreshold = configuration.getInt(STORE_REF_COUNT_THRESHOLD,
+      DEFAULT_STORE_REF_COUNT_THRESHOLD);
+
+  }
+
+  @Override
+  protected void chore() {
+    if (LOG.isTraceEnabled()) {
+      LOG.trace("Starting up Regions Recovery by reopening regions based on storeRefCount...");
+    }
+    try {
+      final ClusterMetrics clusterMetrics = hMaster.getClusterMetrics();
+      final Map<ServerName, ServerMetrics> serverMetricsMap =
+        clusterMetrics.getLiveServerMetrics();
+      final Map<TableName, List<byte[]>> tableToReopenRegionsMap =
+        getTableToRegionsByRefCount(serverMetricsMap);
+      if (MapUtils.isNotEmpty(tableToReopenRegionsMap)) {
+        tableToReopenRegionsMap.forEach((tableName, regionNames) -> {
+          try {
+            LOG.warn("Reopening regions due to high refCount. TableName: {} , noOfRegions: {}",
+              tableName, regionNames.size());
+            hMaster.reopenRegions(tableName, regionNames, NONCE_GENERATOR.getNonceGroup(),
+              NONCE_GENERATOR.newNonce());
+          } catch (IOException e) {
+            LOG.error("{} tableName: {}, regionNames: {}", ERROR_REOPEN_REIONS_MSG,
+              tableName, regionNames, e);
+          }
+        });
+      }
+    } catch (Exception e) {
+      LOG.error("Error while reopening regions based on storeRefCount threshold", e);
+    }
+    if (LOG.isTraceEnabled()) {
+      LOG.trace("Exiting Regions Recovery by reopening regions based on storeRefCount...");
+    }
+  }
+
+  private Map<TableName, List<byte[]>> getTableToRegionsByRefCount(
+      final Map<ServerName, ServerMetrics> serverMetricsMap) {
+
+    final Map<TableName, List<byte[]>> tableToReopenRegionsMap = new HashMap<>();
+    for (ServerMetrics serverMetrics : serverMetricsMap.values()) {
+      Map<byte[], RegionMetrics> regionMetricsMap = serverMetrics.getRegionMetrics();
+      for (RegionMetrics regionMetrics : regionMetricsMap.values()) {
+        final int regionStoreRefCount = regionMetrics.getStoreRefCount();
```

Comment:
This count is the sum of ref count of all HFiles under each of the CF in this region right?  So the sum is not just a factor of many refs on a file. It is a function of number of hfiles under this region. What is it is having so many files at a time and each of that having say <10 active refs. So 250 is obviously not a correct number IMO.  Is there a way we can see refs per HFile?  No need to know ref counts on each of the HFile.  Max number may be enough.  I believe this decision should be made based on the so many ref counts on a file rather than sum of. Thoughts?